### PR TITLE
feat(272): add `chainId` to Keyring API requests (transaction/typed message)

### PR DESCRIPTION
**Pull request by ccharly**
# Description

This is the first step of adding a CAIP-2 scope to internal requests.

It's a "first step" as for now, we cannot get/extract the `chainId` for some methods. This would require a much bigger refactor (which will come later).

## `signTransaction` requests

The `chainId` is part of the transaction request, so we extract it from there and attach it to the internal request.

## `signTypedData` (when applicable) requests

The `chainId` is part of the `domain` field (as defined by EIP-712), however, some older versions of the `eth_signTypedData_vX` might not have it! In this case, we do not provide the `scope`.

As of today, only those methods do not have the `domain`:
- `eth_signTypedData`
- `eth_signTypedData_v1`

# Related issues

- Partially fixes: [MetaMask/accounts-planning#272](https://github.com/MetaMask/accounts-planning/issues/272)

# Testing

### Prerequisites

1. Use the [SSK](https://metamask.github.io/snap-simple-keyring/latest/)
1. Create a snap account (or use an existing one)
1. Enable "Use Synchronous Approval"

### 1. Transaction

1. Send a transaction using your snap account
1. **DO NOT APPROVE YET**
1. List requests using the SSK
1. You should see a scope like: `eip155:${chainId}` (depending on which chain you are using)

### 2. Typed data

1. Use the [E2E dapp](https://metamask.github.io/test-dapp/)
1. Connect to your snap account
1. Sign typed data (using any version you want)
1. **DO NOT APPROVE YET**
1. List requests using the SSK
1. You should see a scope like: `eip155:${chainId}` (depending on which chain you are using)

## Test coverage report

Coverage remains at 100%

File                            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
--------------------------------|---------|----------|---------|---------|-------------------
All files                       |     100 |      100 |     100 |     100 |
 CaseInsensitiveMap.ts          |     100 |      100 |     100 |     100 |
 DeferredPromise.ts             |     100 |      100 |     100 |     100 |
 KeyringSnapControllerClient.ts |     100 |      100 |     100 |     100 |
 SnapIdMap.ts                   |     100 |      100 |     100 |     100 |
 SnapKeyring.ts                 |     100 |      100 |     100 |     100 |
 caip.ts                        |     100 |      100 |     100 |     100 |
 index.ts                       |     100 |      100 |     100 |     100 |
 types.ts                       |     100 |      100 |     100 |     100 |
 util.ts                        |     100 |      100 |     100 |     100 |

## Jest

Added a new test for CAIP helpers:
```console
yarn jest src/caip.test.ts
```

Updated test regarding transaction flow:
```console
yarn jest src/SnapKeyring.test.ts
```

### CODE REVIEW

File: src/caip.test.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    '',
+    'xs',
+    // Not matching
+    '!@#$%^&*()',
+    // Too long
+    'namespacetoolong',
+  ])('throws for invalid namespaces: %s', (namespace) => {
+    const reference = '1';
+    // eslint-disable-next-line jest/require-to-throw-message
+    expect(() => toCaipChainId(namespace, reference)).toThrow();
```
Comment (danroc) [inline src/caip.test.ts:27]: Is there a reason to not include part of the expected error message?

File: src/caip.test.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    // Too short, must have 1 char at least
+    '',
+    // Not matching
+    '!@#$%^&*()',
+    // Too long
+    '012345678901234567890123456789012', // 33 chars
+  ])('throws for invalid reference: %s', (reference) => {
+    const namespace = CaipNamespaces.Eip155;
+    // eslint-disable-next-line jest/require-to-throw-message
+    expect(() => toCaipChainId(namespace, reference)).toThrow();
```
Comment (danroc) [inline src/caip.test.ts:40]: Same as above

File: src/SnapKeyring.test.ts [hunk 3]
```diff
@@ -706,6 +734,53 @@ describe('SnapKeyring', () => {
...
+      });
+
+      const dataToSignWithoutDomainChainId = {
+        ...dataToSign,
+        domain: {
+          name: dataToSign.domain.name,
+          version: dataToSign.domain.version,
+          verifyingContract: dataToSign.domain.verifyingContract,
+          // We do not defined the chainId (it's currently marked as
+          // optional in the current type declation).
```
Comment (danroc) [inline src/SnapKeyring.test.ts:750]: Typo: `declation`

File: src/caip.test.ts (hunk 1 above)
Comment (ccharly) [reply src/caip.test.ts:27]: Is being lazy a good reason? :trollface: You're right, I'm gonna fix that right now! :)

File: src/SnapKeyring.test.ts (hunk 3 above)
Comment (ccharly) [reply src/SnapKeyring.test.ts]: Fixed!

File: src/caip.test.ts (hunk 1 above)
Comment (ccharly) [reply src/caip.test.ts]: Fixed!

**Review by danroc — APPROVED** [review]

**Pull request merged**